src/api/alchemy-wallet.ts

```diff
@@ -1,3 +1,105 @@
+export class Wallet extends EthersWallet {
+  readonly hasAlchemyObject: boolean = false;
+  readonly alchemy?: Alchemy;
+  alchemyProvider?: Provider;
 
-export class Wallet extends EthersWallet {}
+  constructor(
+    privateKey: BytesLike | ExternallyOwnedAccount | SigningKey,
+    alchemyOrProvider?: Alchemy | Provider
+  ) {
+    // If object passed in is a provider, send to super
+    let superProvider: Provider | undefined;
+    if (alchemyOrProvider && Provider.isProvider(alchemyOrProvider)) {
+      superProvider = alchemyOrProvider;
+    }
+    super(privateKey, superProvider);
+
+    // If object passed in is an Alchemy object, just set Alchemy
+    if (alchemyOrProvider && !Provider.isProvider(alchemyOrProvider)) {
+      this.hasAlchemyObject = true;
+      this.alchemy = alchemyOrProvider;
+    }
+  }
+
+  //////////////////////////////////////////////////////////////////
+  // Set of overrides from Signer to handle async provider retrieval.
+  //////////////////////////////////////////////////////////////////
+
+  async getBalance(blockTag?: BlockTag): Promise<BigNumber> {
+    return this.getWallet().then(wallet => wallet.getBalance(blockTag));
+  }
+
+  async getTransactionCount(blockTag?: BlockTag): Promise<number> {
+    return this.getWallet().then(wallet =>
+      wallet.getTransactionCount(blockTag)
+    );
+  }
+
+  // Populates "from" if unspecified, and estimates the gas for the transaction
+  async estimateGas(
+    transaction: Deferrable<TransactionRequest>
+  ): Promise<BigNumber> {
+    return this.getWallet().then(wallet => wallet.estimateGas(transaction));
+  }
+
+  // Populates "from" if unspecified, and calls with the transaction
+  async call(
+    transaction: Deferrable<TransactionRequest>,
+    blockTag?: BlockTag
+  ): Promise<string> {
+    return this.getWallet().then(wallet => wallet.call(transaction, blockTag));
+  }
+
+  // Populates all fields in a transaction, signs it and sends it to the network
+  async sendTransaction(
+    transaction: Deferrable<TransactionRequest>
+  ): Promise<TransactionResponse> {
+    return this.getWallet().then(wallet => wallet.sendTransaction(transaction));
+  }
+
+  async getChainId(): Promise<number> {
+    return this.getWallet().then(wallet => wallet.getChainId());
+  }
+
+  async getGasPrice(): Promise<BigNumber> {
+    return this.getWallet().then(wallet => wallet.getGasPrice());
+  }
+
+  async getFeeData(): Promise<FeeData> {
+    return this.getWallet().then(wallet => wallet.getFeeData());
+  }
+
+  async resolveName(name: string): Promise<string> {
+    return this.getWallet().then(wallet => wallet.resolveName(name));
+  }
+
+  private async getWallet() {
+    if (!this.hasAlchemyObject) {
+      return this.connect(this.provider);
+    }
+    if (!this.alchemy) {
+      throw new Error('No Alchemy object passed into the wallet.');
+    }
+    if (!this.alchemyProvider) {
+      this.alchemyProvider = await this.alchemy.config.getProvider();
+    }
```

Comment:
This could be improved: as written, if the user makes multiple calls at once, then we could end up calling `getProvider()` multiple times, which can have a real cost associated with if if the provider is a WebSocket provider because it will open multiple WebSocket connections that will never be closed.

Instead, consider structuring this class as follows, keeping a single promise around so that multiple calls before the provider is created end up getting the same provider:
```ts
export class Wallet extends EthersWallet {
  private alchemyProviderPromise?: Promise<Provider>;

  // ...

  private async getWallet(): Promise<Provider> {
    // ...
    if (!this.alchemyProviderPromise) {
      this.alchemyProviderPromise = this.alchemy.config.getProvider();
    }
    return this.connect(await this.alchemyProviderPromise);
  }
}
```